### PR TITLE
Week 14 - Karin Nordkvist (Todo)

### PR DESCRIPTION
**Pull request by karinnordkvist**
**Link to my deployed project:** https://todos-by-karin.netlify.app/ ✏️📋

**My reflections on how this week's project turned out:**
I'm pleased with the result of this week's project. I felt like I learned a lot, and that I was able to try out a lot of things. Super fun!

**Things I'd like to have clarified or explained in more detail:**
Working with immutability in reducer actions. 💀

### CODE REVIEW

File: code/public/index.html [hunk 1]
```diff
@@ -1,34 +1,18 @@
...
-      Notice the use of %PUBLIC_URL% in the tags above.
-      It will be replaced with the URL of the `public` folder during the build.
-      Only files inside the `public` folder can be referenced from the HTML.
-
-      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
-      work correctly both with client-side routing and a non-root public URL.
-      Learn how to configure a non-root public URL by running `npm run build`.
-    -->
-    <title>Technigo React App</title>
+    <link
```
Comment (aagneshansson) [inline code/public/index.html:6]: Really nice favicon!

File: code/src/App.js [hunk 2]
```diff
@@ -1,9 +1,55 @@
...
+import React, { useState } from 'react';
+import { Provider } from 'react-redux';
+import { combineReducers, createStore } from '@reduxjs/toolkit';
+
+// Reducers
+import { todos } from './reducers/todos';
+
+// Components
+import { ListOfTodos } from './components/ListOfTodos';
+import { Header } from './components/Header';
```
Comment (aagneshansson) [inline code/src/App.js:10]: Very nice structure.

File: code/src/components/CounterAndSorter.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
+import styled from 'styled-components';
+import Select from 'react-select';
+
+// Styling
+import { InnerFlexWrapper, InfoText } from '../styling/GlobalStyling';
```
Comment (aagneshansson) [inline code/src/components/CounterAndSorter.js:7]: Great that you have worked with global styling 🌟

File: code/src/components/Footer.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import styled from 'styled-components';
+
+// -----------------------------------------------------------------------------
+
+export const Footer = () => {
+  // Sneaky hidden function to clear local storage
```
Comment (aagneshansson) [inline code/src/components/Footer.js:7]: Awesome done! Will use this in my own code 😃

File: code/src/App.js [hunk 5]
```diff
@@ -1,9 +1,55 @@
...
+// New storage
+const persistedStateJSON = localStorage.getItem('reduxState');
+let persistedState = {};
+
+if (persistedStateJSON) {
+  persistedState = JSON.parse(persistedStateJSON);
+}
+
+const store = createStore(
+  reducer,
```
Comment (aagneshansson) [inline code/src/App.js:28]: It's really impressive you stretched your goals and created local storage 👏

File: code/src/components/CounterAndSorter.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+  // Count how many incomplete tasks there are in selected category
+  const amountOfItems = (category) => {
+    return nonCompletedTasks.filter((task) => task.category === category)
+      .length;
+  };
+
+  // Category options
+  const options = [
+    { value: 'All', label: `All (${nonCompletedTasks.length})` },
```
Comment (aagneshansson) [inline code/src/components/CounterAndSorter.js:31]: This was related to the package you installed, right? Would be great to learn more about that 👍

File: code/src/components/CounterAndSorter.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+const CustomSelect = styled(Select)`
+  width: 240px;
+  margin-right: 20px;
+`;
+
+const InnerFlexWrapperColumn = styled(InnerFlexWrapper)`
+  flex-direction: column;
+  align-items: flex-start;
+`;
+
```
Comment (aagneshansson) [inline code/src/components/CounterAndSorter.js:78]: Is there a technical different between doing "styled(InnerFlexWrapper)" and "styled.div"? Or why have you done differently in different places? 😄

File: code/src/components/CustomCheckbox.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+`;
+
+const CheckedSign = styled.p`
+  color: ${(props) => (props.checked ? 'white' : '#4300ca')};
+`;
+
+const StyledCheckbox = styled.div`
+  display: inline-flex;
+  align-items: center;
+  justify-content: center;
```
Comment (aagneshansson) [inline code/src/components/CustomCheckbox.js:59]: Love that you did styled checkboxes, know how tricky it was 😆

File: code/src/components/CustomCheckbox.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  width: 30px;
+  height: 30px;
+  background: ${(props) => (props.checked ? '#4300ca' : '#f5f1fc')};
+  border-radius: 50%;
+
+  &:hover {
```
Comment (aagneshansson) [inline code/src/components/CustomCheckbox.js:65]: Really nice styling!

File: code/src/components/CustomCheckbox.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  color: ${(props) => (props.checked ? 'white' : '#4300ca')};
+`;
+
+const StyledCheckbox = styled.div`
+  display: inline-flex;
+  align-items: center;
+  justify-content: center;
+  width: 30px;
+  height: 30px;
+  background: ${(props) => (props.checked ? '#4300ca' : '#f5f1fc')};
```
Comment (aagneshansson) [inline code/src/components/CustomCheckbox.js:62]: This.is.just.great!

File: README.md [hunk 11]
```diff
@@ -1,13 +1,19 @@
...
 
-Start by briefly describing the assignment in a sentence or two. Keep it short and to the point.
+Turns out I really like working with Redux, and even though many things were new to me, I feel like I was able to try all sorts of things out. By the end of the project I could focus more on zooming in on each action, making sure I understand exatcly what I have done and why. Immutability is still a bit fuzzy for me, though. I found this project to be very valuable regarding handling arrays, using map, filter and find a lot.
 
-## The problem
+If I'd have more time, I would like to look into how to work with custom alerts, reminders and setting deadlines using dates, also being able to export/import tasks from/to desktop/mobile etc. Maybe it would have been nice practice to let the user add its own categories, and by doing so also updating the filter by category-functions.
```
Comment (aagneshansson) [inline README.md:9]: Karin, you should be SO proud, don't have really much more to say than you have done an amazing job! Your code is clean and simple to follow, you have stretched the goals to black level and styled it so so nice 👏 🏅 😄 

Very well done! Continue like this 👍 (I am gonna use this app) ❤️

**Review by aagneshansson — APPROVED** [review]
VERY well done Karin!!!

File: code/src/components/CounterAndSorter.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    <CounterAndSorterWrapper>
+      <InnerFlexWrapperColumn>
+        <InfoText>Sort by category:</InfoText>
+        <CustomSelect
+          value={showingCategory.value}
+          onChange={handleChange}
+          options={options}
+          placeholder="Showing All"
+          className="react-select-container"
+        />
```
Comment (HIPPIEKICK) [inline code/src/components/CounterAndSorter.js:54]: Love this!

File: code/src/components/CounterAndSorter.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+      </InnerFlexWrapperColumn>
+      <InfoTextCounter>
+        Completed: {completedTasks.length} / {allTodos.length}
+      </InfoTextCounter>
+    </CounterAndSorterWrapper>
+  );
+};
+
+// Local styles -----------------------------
```
Comment (HIPPIEKICK) [inline code/src/components/CounterAndSorter.js:63]: You can cretae more than one style file if you'd like!

File: code/src/components/Footer.js (hunk 4 above)
Comment (HIPPIEKICK) [reply code/src/components/Footer.js:7]: Haha this caused me to accidentally remove my todos 🙈

File: code/src/components/ListOfTodos.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+          <ControllerWrapper>
+            <p>Remove all tasks</p>
+            <Button onClick={handleRemoveAll}>✕</Button>
+          </ControllerWrapper>
+
+          <ControllerWrapper>
+            <p>Remove all completed tasks</p>
+            <Button onClick={handleRemoveCompleted}>✕</Button>
+          </ControllerWrapper>
```
Comment (HIPPIEKICK) [inline code/src/components/ListOfTodos.js:114]: Would be cool if this was only rendered if there are any completed tasks 😎

Comment (HIPPIEKICK) [reply code/src/components/ListOfTodos.js:114]: And same with the other buttons, with their respective logic

File: code/src/reducers/todos.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+export const todos = createSlice({
+  name: 'todos',
+  initialState: initialState,
+  reducers: {
+    addTodo: (state, action) => {
+      const newTodosList = [...state.tasks, action.payload];
+      state.tasks = newTodosList;
+    },
+    checkTodo: (state, action) => {
```
Comment (HIPPIEKICK) [inline code/src/reducers/todos.js:16]: This can also uncheck todos 😎

File: code/src/reducers/todos.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+      const removedAllTodos = [];
+      state.tasks = removedAllTodos;
+    },
+    removeAllCompletedTodos: (state, action) => {
+      const removedCompletedTodos = state.tasks.filter(
+        (task) => !task.isCompleted
+      );
+      state.tasks = removedCompletedTodos;
+    },
+  },
```
Comment (HIPPIEKICK) [inline code/src/reducers/todos.js:43]: Some excessive commas

**Pull request closed without merging**